### PR TITLE
Add support for multiple BWC Versions

### PR DESCRIPTION
**Pull request by naveentatikonda**
Signed-off-by: Naveen Tatikonda <navtat@amazon.com>

### Description
Improve BWC framework and add support for multiple BWC Versions
 
### Issues Resolved
https://github.com/opensearch-project/k-NN/issues/360
 
### Check List
- [x] All tests pass
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: .github/workflows/backwards_compatibility_tests_workflow.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+      - "*"
+      - "feature/**"
+
+jobs:
+  Restart-Upgrade-BWCTests-k-NN:
+    strategy:
+      matrix:
+        java: [ 11, 17 ]
+        bwc_version : [ "1.0.0", "1.1.0", "1.2.4", "1.3.1" ]
+        opensearch_version : [ "2.0.0-rc1-SNAPSHOT" ]
```
Comment (martin-gaievski) [inline .github/workflows/backwards_compatibility_tests_workflow.yml:18]: can we define multiple values for opensearch_version?

**Review by martin-gaievski — DISMISSED** [review]

File: .github/workflows/backwards_compatibility_tests_workflow.yml (hunk 1 above)
Comment (naveentatikonda) [reply .github/workflows/backwards_compatibility_tests_workflow.yml:18]: yes, we can like we did for bwc_version

File: .github/workflows/backwards_compatibility_tests_workflow.yml [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    branches:
+      - "*"
+      - "feature/**"
+
+jobs:
+  Restart-Upgrade-BWCTests-k-NN:
+    strategy:
+      matrix:
+        java: [ 11, 17 ]
+        bwc_version : [ "1.0.0", "1.1.0", "1.2.4", "1.3.1" ]
```
Comment (jmazanec15) [inline .github/workflows/backwards_compatibility_tests_workflow.yml:17]: 1.3.2 as of today

File: qa/restart-upgrade/build.gradle [hunk 3]
```diff
@@ -25,34 +26,44 @@ testClusters {
...
-    systemProperty 'tests.security.manager', 'false'
-}
+    task testAgainstOldCluster(type: StandaloneRestIntegTestTask) {
+        dependsOn "zipBwcPlugin"
+        useCluster testClusters."${baseName}"
+        systemProperty 'tests.rest.bwcsuite_cluster', 'old_cluster'
+        systemProperty 'tests.is_old_cluster', 'true'
+        systemProperty 'tests.skip_delete_model_index', 'true'
+        systemProperty 'tests.plugin_bwc_version', knn_bwc_version
+        if (knn_bwc_version.startsWith("1.0") || knn_bwc_version.startsWith("1.1")) {
```
Comment (jmazanec15) [inline qa/restart-upgrade/build.gradle:36]: Nice, this makes code much better. For filtering out tests, add a comment explaining why they are filtered out.

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/IndexingIT.java [hunk 4]
```diff
@@ -129,9 +135,9 @@ public void testNullParametersOnUpgrade() throws Exception {
     // test empty parameters
     public void testEmptyParametersOnUpgrade() throws Exception {
 
-        // Skip test if version is 1.2 or 1.3
+        // Skip test if version is 1.x.y
```
Comment (jmazanec15) [inline qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/IndexingIT.java:138]: We actually do want to test 1.0 and 1.1

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/WarmupIT.java [hunk 5]
```diff
@@ -72,6 +73,12 @@ public void testKNNWarmupDefaultMethodFieldMapping() throws Exception {
     // Custom Method Field Mapping
     // space_type : "innerproduct", engine : "faiss", m : 50, ef_construction : 1024
     public void testKNNWarmupCustomMethodFieldMapping() throws Exception {
+
+        // Skip test if version is 1.0 or 1.1
+        Optional<String> bwcVersion = getBWCVersion();
+        if (bwcVersion.isEmpty() || bwcVersion.get().startsWith("1.0") || bwcVersion.get().startsWith("1.1")) {
```
Comment (jmazanec15) [inline qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/WarmupIT.java:79]: Why not move this to the gradle to get rid of if/else statement?

File: qa/rolling-upgrade/build.gradle [hunk 6]
```diff
@@ -9,12 +9,13 @@ apply from : "$rootDir/qa/build.gradle"
 
 String knn_bwc_version = System.getProperty("bwc.version")
 String baseName = "knnBwcCluster-rolling"
+String opensearch_version_upgraded = System.getProperty("rolling.bwctests.opensearch.version", opensearch_version)
```
Comment (jmazanec15) [inline qa/rolling-upgrade/build.gradle:12]: Is it actually possible to specify a different version than current to upgrade to?

File: .github/workflows/backwards_compatibility_tests_workflow.yml (hunk 2 above)
Comment (naveentatikonda) [reply .github/workflows/backwards_compatibility_tests_workflow.yml]: updated

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/IndexingIT.java (hunk 4 above)
Comment (naveentatikonda) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/IndexingIT.java]: This test is not compatible with 1.0, 1.1, 1.2 or 1.3. So, I'm skipping this test if the version starts with "1."

File: qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/WarmupIT.java (hunk 5 above)
Comment (naveentatikonda) [reply qa/restart-upgrade/src/test/java/org/opensearch/knn/bwc/WarmupIT.java]: updated

File: qa/rolling-upgrade/build.gradle (hunk 6 above)
Comment (naveentatikonda) [reply qa/rolling-upgrade/build.gradle:12]: yes, we need to add it to that matrix

**Review by martin-gaievski — APPROVED** [review]
Looks good, thank you

File: qa/restart-upgrade/build.gradle [hunk 7]
```diff
@@ -19,40 +20,86 @@ testClusters {
         plugin(project.tasks.zipBwcPlugin.archiveFile)
         setting 'path.repo', "${buildDir}/cluster/shared/repo/${baseName}"
         setting 'http.content_type.required', 'true'
-        environment "LD_LIBRARY_PATH", "${buildDir}/testclusters/${baseName}-0/distro/${knn_bwc_version}-ARCHIVE/plugins/opensearch-knn/knnlib"
-        systemProperty "java.library.path", "${buildDir}/testclusters/${baseName}-0/distro/${knn_bwc_version}-ARCHIVE/plugins/opensearch-knn/knnlib"
+        environment "LD_LIBRARY_PATH", "${buildDir}/testclusters/${baseName}-0/distro/${knn_bwc_version}-ARCHIVE/plugins/opensearch-knn/knnlib;${buildDir}/testclusters/${baseName}-0/distro/${knn_bwc_version}-ARCHIVE/plugins/opensearch-knn/lib"
```
Comment (jmazanec15) [inline qa/restart-upgrade/build.gradle:23]: This will fail for 1.0 and 1.1 and 1.2?

Comment (naveentatikonda) [reply qa/restart-upgrade/build.gradle:23]: Do you mean when the version is upgraded to 1.0, 1.1 or 1.2?
Or upgraded from these versions to latest OpenSearch version?

Comment (jmazanec15) [reply qa/restart-upgrade/build.gradle:23]: Upgrade from

Comment (naveentatikonda) [reply qa/restart-upgrade/build.gradle:23]: no, they don't fail. We can verify it from the checks. The checks cover all those scenarios and all the checks have passed.

Comment (jmazanec15) [reply qa/restart-upgrade/build.gradle:23]: Oh I see so you set the location in both places. I think that makes sense.

**Review by jmazanec15 — APPROVED** [review]

**Pull request merged**